craft_application/launchpad/models/recipe.py

```diff
@@ -0,0 +1,332 @@
+
+    @staticmethod
+    def _fill_repo_info(
+        kwargs: dict[str, Any],
+        *,
+        git_ref: str | None = None,
+        bzr_branch: str | None = None,
+    ) -> None:
+        """Conditionally fill source repository info into keyword arguments."""
+        if (git_ref and bzr_branch) or (not git_ref and not bzr_branch):
+            raise ValueError(
+                "A snap recipe must refer to a git repository or a bazaar branch, "
+                "but not both.",
+            )
+        if git_ref:
+            kwargs["git_ref"] = git_ref
+        if bzr_branch:
+            kwargs["branch"] = bzr_branch
+
+    def get_builds(self) -> Collection[build.Build]:
+        """Get the existing builds for a Recipe."""
+        return [build.Build(self._lp, b) for b in self._obj.builds]
+
+    def build(
+        self,
+        archive: str = "/ubuntu/+archive/primary",
+        pocket: Pocket = Pocket.UPDATES,
+        channels: BuildChannels | None = None,
+        deadline: int | None = None,
+    ) -> Collection[build.Build]:
+        """Create a new set of builds for this recipe."""
+        request_build_kwargs = {
+            "archive": archive,
+            "pocket": pocket.value,
+        }
+        if channels:
+            request_build_kwargs["channels"] = channels
+        build_request = self._obj.requestBuilds(**request_build_kwargs)
+        sleep_time = 0.5
+        while build_request.status == "Pending":
+            # Check to see if we've run out of time.
+            if deadline is not None and time.monotonic_ns() >= deadline:
+                raise TimeoutError
+            time.sleep(sleep_time)
+            sleep_time *= 1.1
+            build_request.lp_refresh()
+
+        if build_request.status != "Completed":
+            raise errors.BuildError("Build request failed")
+
+        return [build.Build(self._lp, obj) for obj in build_request.builds]
+
+
+class _StoreRecipe(_BaseRecipe):
+    """A recipe for an item that has a store entry."""
+
+    store_name: str
+    """The name of the package in the store."""
+
+    @staticmethod
+    def _fill_store_info(
+        kwargs: dict[str, Any],
+        *,
+        store_name: str | None,
+        store_channels: Collection[str],
+    ) -> None:
+        """Conditionally fill store info into store-related keyword arguments."""
+        if store_name:
+            kwargs["store_name"] = store_name
+            kwargs["store_channels"] = store_channels
+
+
+class SnapRecipe(_StoreRecipe):
+    """A recipe for a snap.
+
+    https://api.launchpad.net/devel.html#snap
+    """
+
+    @classmethod
+    @override
+    def new(
+        cls,
+        lp: Launchpad,
+        name: str,
+        owner: str,
+        *,
+        architectures: Collection[str] | None = None,
+        description: str | None = None,
+        project: str | None = None,
+        # Repository. Must be either a git repository or a Bazaar branch but not both.
```

Comment:
Would prefer docstrings but I'm OK with this style.

Reply:
I've added some docstrings but kept these comments as it helps developers group the large number of parameters.